File: src/cmr/spatial/orbits/orbit_geometry.clj
(ns cmr.spatial.orbits.orbit-geometry
  "Contains functions for converting orbit parameters into polygons"
  (:require [cmr.spatial.point :as p]
            [cmr.spatial.math :refer :all]
            [cmr.common.services.errors :as errors]
            [primitive-math]
            [cmr.spatial.mbr :as m]
            [cmr.spatial.conversion :as c]
            [cmr.spatial.ring-relations :as rr]
            [cmr.spatial.arc :as a]
            [cmr.spatial.derived :as d]
            [cmr.spatial.polygon :as poly]
            [cmr.spatial.orbits.echo-orbits :as echo-orbits]
            [cmr.spatial.orbits.coordinate :as coordinate]))

(primitive-math/use-primitive-operators)

;; Initial implementation here is a complete straightforward reimplementation of ruby to clojure

;; TODO rename and refactor as appropriate.
;; - some of this code should be moved into the indexer



(defn safe-asin
  ^double [^double v]
  (asin (if (>= v 0.0)
          (min v 1.0)
          (max v -1.0))))

(defn safe-acos
  ^double [^double v]
  (acos (if (>= v 0.0)
          (min v 1.0)
          (max v -1.0))))

(defn longitude-correction
  ^double [^double time-elapsed-secs]
  (mod (* (/ TAU echo-orbits/SOLAR_DAY_S) time-elapsed-secs) TAU))

(defn ground-track-uncorrected
  [orbit-parameters ^double ascending-crossing-lon-rad ^double alpha]
  (let [inclination-rad (radians (:inclination-angle orbit-parameters))
        ground-track-lat (asin (* (sin inclination-rad) (sin alpha)))
        ground-track-lon (if (= (abs ground-track-lat) (/ PI 2.0))
                           ascending-crossing-lon-rad
                           (let [drift (safe-acos (/ (cos alpha) (cos ground-track-lat)))]
                             (- ascending-crossing-lon-rad
                                (* (if (<= alpha PI)
                                     drift
                                     (- TAU drift))
                                   (if (echo-orbits/retrograde? orbit-parameters)
                                     1.0
                                     -1.0)))))]
    (coordinate/from-phi-theta ground-track-lat ground-track-lon)))


(def ^:const ^double ALPHA_CORRECTION_DELTA
  "A small number to offset alpha to keep it from being exactly 0 or PI"
  0.00000001)

(defn alpha-negate
  "Multiplies v * -1 if alpha is greater than pi"
  [^double alpha ^double v]
  (if (> alpha PI)
    (* v -1.0)
    v))

(def ^:const ^double HALF_PI
  (/ PI 2.0))

;; TODO rename this
(defn along-track-swath-edges
  "TODO"
  [orbit-parameters ^double ascending-crossing-lon ^double time-elapsed-mins]
  (let [inclination-rad (echo-orbits/inclination-rad orbit-parameters)
        time-elapsed-secs (* time-elapsed-mins 60.0)
        ascending-crossing-lon-rad (radians ascending-crossing-lon)

        ^double alpha (mod (* time-elapsed-secs (echo-orbits/angular-velocity-rad-s orbit-parameters)) TAU)
        ^double alpha (if (or (= 0.0 alpha) (= PI alpha))
                        (+ alpha ALPHA_CORRECTION_DELTA)
                        alpha)

        r (/ (echo-orbits/swath-width-rad orbit-parameters) 2.0)
        coord (ground-track-uncorrected orbit-parameters 0 alpha)
        beta (acos (* (cos r) (cos (:phi coord)) (cos (:theta coord))))
        rR (safe-asin (/ (sin r) (sin beta)))

        lat-left (alpha-negate alpha (asin (* (sin (+ rR inclination-rad)) (sin beta))))
        rw (safe-acos (* (cos r) (cos (:phi coord)) (/ (cos (:theta coord)) (cos lat-left))))
        lon-left (- ascending-crossing-lon-rad
                    (* (if (<= alpha PI) rw (- TAU rw))
                       (if (< (+ rR inclination-rad) HALF_PI) -1.0 1.0)))

        lat-right (alpha-negate alpha (asin (* (sin (+ (* -1.0 rR) inclination-rad)) (sin beta))))
        re (safe-acos (* (cos r) (cos (:phi coord)) (/ (cos (:theta coord)) (cos lat-right))))
        lon-right (- ascending-crossing-lon-rad (* (if (<= alpha PI) re (- TAU re))
                                                   (if (< (- inclination-rad rR) HALF_PI) -1.0 1.0)))

        lon-correct (longitude-correction time-elapsed-secs)
        edge [(coordinate/from-phi-theta lat-left (- lon-left lon-correct))
              (coordinate/from-phi-theta lat-right (- lon-right lon-correct))]]
    (if (< alpha PI)
      edge
      (reverse edge))))

(comment

  ;; an example from orbit geometry spec
  (def test-orbit-parameters
    {:inclination-angle 120.0
     :period 100.0
     :swath-width 1450.0
     :start-circular-latitude -90.0
     :number-of-orbits 0.5})

  (for [{:keys [lon lat]} (along-track-swath-edges test-orbit-parameters 50.0 10)]
    [lon lat])

  (ground-track
    test-orbit-parameters
    0.0
    30.0)


  )

(defn ground-track
  "TODO"
  [orbit-parameters ^double ascending-crossing-lon ^double time-elapsed-mins]
  (let [time-elapsed-secs (* 60.0 time-elapsed-mins)
        ascending-crossing-lon-rad (radians ascending-crossing-lon)
        alpha (mod (* time-elapsed-secs (echo-orbits/angular-velocity-rad-s orbit-parameters)) TAU)
        coord (ground-track-uncorrected orbit-parameters ascending-crossing-lon-rad alpha)]
    (coordinate/from-phi-theta (:phi coord) (- ^double (:theta coord)
                                               (longitude-correction time-elapsed-secs)))))

File: orbits-lib/resources/orbits/orbit_geometry.rb

require 'orbits/coordinate'



module Orbits
  class OrbitGeometry

  	include Math

  	DELTA = 0.00000001
  	attr_reader :orbit

	def initialize(orbit)
      @orbit = orbit
	end

	def ground_track(equator_ascending_crossing_longitude_deg, time_elapsed_min)
      time_elapsed_s = time_elapsed_min * 60
      equator_ascending_crossing_longitude_rad = equator_ascending_crossing_longitude_deg * PI/180
      coord = ground_track_uncorrected(equator_ascending_crossing_longitude_rad, (time_elapsed_s * orbit.angular_velocity_rad_s) % (2 * PI))
      Coordinate.phi_theta(coord.phi, coord.theta - longitude_correction(time_elapsed_s))
    end

    def along_track_swath_edges(equator_ascending_crossing_longitude_deg, time_elapsed_min)
      time_elapsed_s = time_elapsed_min * 60
      equator_ascending_crossing_longitude_rad = equator_ascending_crossing_longitude_deg * PI/180
      alpha = (time_elapsed_s * orbit.angular_velocity_rad_s) % (2 * PI)
      alpha = alpha + DELTA if alpha == 0 || alpha == PI

      r = orbit.swath_width_rad/2
      coord = ground_track_uncorrected(0, alpha)
      beta = Math::acos(cos(r) * cos(coord.phi) * cos(coord.theta))
      rR = asin(sin(r)/sin(beta))

      lat_left =   Math::asin(sin(rR + orbit.inclination_rad) * sin(beta)) * ((alpha <= PI) ? 1 : -1)
      rw = acos(cos(r) * cos(coord.phi) * cos(coord.theta)/cos(lat_left))
      lon_left = equator_ascending_crossing_longitude_rad - (alpha <= PI ? rw : 2 * PI - rw) * (rR + orbit.inclination_rad < PI/2 ? -1 : 1)

      lat_right = Math::asin(sin(-rR + orbit.inclination_rad) * sin(beta)) * ((alpha <= PI) ? 1 : -1)
      re = acos(cos(r) * cos(coord.phi) * cos(coord.theta)/cos(lat_right))
      lon_right = equator_ascending_crossing_longitude_rad -  (alpha <= PI ? re : 2 * PI - re) * (-rR + orbit.inclination_rad < PI/2 ? -1 : 1) 

      edge = [ Coordinate.phi_theta(lat_left, lon_left - longitude_correction(time_elapsed_s)),  Coordinate.phi_theta(lat_right, lon_right - longitude_correction(time_elapsed_s))]
      
      if(alpha < PI)
      	edge
      else
      	edge.reverse
      end

    end

    private

    def ground_track_uncorrected(equator_ascending_crossing_longitude_rad, alpha)
      ground_track_latitude = Math.asin(sin(orbit.inclination_rad) * sin(alpha))
      
      if(ground_track_latitude.abs == PI/2)
        ground_track_longitude = equator_ascending_crossing_longitude_rad
      else
      	drift = acos(cos(alpha)/cos(ground_track_latitude))
        ground_track_longitude = equator_ascending_crossing_longitude_rad -  (alpha <= PI ? drift : 2 * PI - drift) * (orbit.retrograde? ? 1 : -1)
      end
      Coordinate.phi_theta(ground_track_latitude, ground_track_longitude)
    end

    def longitude_correction(time_elapsed_s)
      (2 * PI/Orbits::SOLAR_DAY_S) * time_elapsed_s % (2*Math::PI)
    end

    def acos(val)
      theta = Math::acos(val >=0 ? [val, 1].min : [val, -1].max)
    end

    def asin(val)
      theta = Math::asin(val >=0 ? [val, 1].min : [val, -1].max)
    end

  end
end